core/src/main/scala/cats/Monad.scala

```diff
@@ -14,4 +14,90 @@ import simulacrum.typeclass
 @typeclass trait Monad[F[_]] extends FlatMap[F] with Applicative[F] {
   override def map[A, B](fa: F[A])(f: A => B): F[B] =
     flatMap(fa)(a => pure(f(a)))
+
+  /**
+   * Execute an action repeatedly as long as the given `Boolean` expression
+   * returns `true`. The condition is evaluated before the loop body.
+   * Collects the results into an arbitrary `Alternative` value, such as a `Vector`.
+   * This implementation uses append on each evaluation result,
+   * so avoid data structures with non-constant append performance, e.g. `List`.
+  */
+  def whileM[G[_], A](p: F[Boolean])(body: => F[A])(implicit G: Alternative[G]): F[G[A]] = {
+    val b = Eval.later(body)
+    tailRecM[G[A], G[A]](G.empty)(xs => ifM(p)(
+      ifTrue = {
+        map(b.value) { bv =>
+          Left(G.combineK(xs, G.pure(bv)))
+        }
+      },
+      ifFalse = pure(xs.asRight[G[A]])
+    ))
+  }
+
+  /**
+   * Execute an action repeatedly as long as the given `Boolean` expression
+   * returns `true`. The condition is evaluated before the loop body.
+   * Discards results.
+   */
+  def whileM_[A](p: F[Boolean])(body: => F[A]): F[Unit] = {
+    val continue: Either[Unit, Unit] = Left(())
+    val stop: F[Either[Unit, Unit]] = pure(Right(()))
+    val b = Eval.later(body)
```

Comment:
Do you think these methods would be of practical use for `Future` or `Try`? Neither a `Try[Boolean]` nor `Future[Boolean]` will ever change the underneath value, and they are call by value.
It seems to me that the original incentive of introducing these methods are **only** for data types that track side effects such as `IO`/`Task`/`StateT`. When it comes to other data types like `Option` or `List`, they don't make sense to me, at least with current API design.